db/python/tables/bq/generic_bq_filter.py

```diff
@@ -0,0 +1,101 @@
+    ) -> tuple[str, dict[str, T | list[T] | Any | list[Any]]]:
+        """
+        Convert to SQL, and avoid SQL injection
+
+        """
+        conditionals = []
+        values: dict[str, T | list[T] | Any | list[Any]] = {}
+        _column_name = column_name or column
+
+        if not isinstance(column, str):
+            raise ValueError(f'Column {_column_name!r} must be a string')
+        if self.eq is not None:
+            k = self.generate_field_name(_column_name + '_eq')
+            conditionals.append(f'{column} = {self._sql_cond_prep(k, self.eq)}')
+            values[k] = self._sql_value_prep(k, self.eq)
+        if self.in_ is not None:
+            if not isinstance(self.in_, list):
+                raise ValueError('IN filter must be a list')
+            if len(self.in_) == 1:
+                k = self.generate_field_name(_column_name + '_in_eq')
+                conditionals.append(f'{column} = {self._sql_cond_prep(k, self.in_[0])}')
+                values[k] = self._sql_value_prep(k, self.in_[0])
+            else:
+                k = self.generate_field_name(_column_name + '_in')
+                conditionals.append(f'{column} IN ({self._sql_cond_prep(k, self.in_)})')
+                values[k] = self._sql_value_prep(k, self.in_)
+        if self.nin is not None:
+            if not isinstance(self.nin, list):
+                raise ValueError('NIN filter must be a list')
+            k = self.generate_field_name(column + '_nin')
+            conditionals.append(f'{column} NOT IN ({self._sql_cond_prep(k, self.nin)})')
+            values[k] = self._sql_value_prep(k, self.nin)
+        if self.gt is not None:
+            k = self.generate_field_name(column + '_gt')
+            conditionals.append(f'{column} > {self._sql_cond_prep(k, self.gt)}')
+            values[k] = self._sql_value_prep(k, self.gt)
+        if self.gte is not None:
+            k = self.generate_field_name(column + '_gte')
+            conditionals.append(f'{column} >= {self._sql_cond_prep(k, self.gte)}')
+            values[k] = self._sql_value_prep(k, self.gte)
+        if self.lt is not None:
+            k = self.generate_field_name(column + '_lt')
+            conditionals.append(f'{column} < {self._sql_cond_prep(k, self.lt)}')
+            values[k] = self._sql_value_prep(k, self.lt)
+        if self.lte is not None:
+            k = self.generate_field_name(column + '_lte')
+            conditionals.append(f'{column} <= {self._sql_cond_prep(k, self.lte)}')
+            values[k] = self._sql_value_prep(k, self.lte)
+
+        return ' AND '.join(conditionals), values
+
+    @staticmethod
+    def _sql_cond_prep(key, value) -> str:
+        """
+        By default '@{key}' is used,
+        but for datetime it has to be wrapped in TIMESTAMP(@{k})
+        """
+        if isinstance(value, datetime):
+            return f'TIMESTAMP(@{key})'
+
+        # otherwise as default
+        return f'@{key}'
```

Comment:
Is this key sanitised? If no, can we do something like replace anything `[^A-Za-z0-9_]` with an underscore?

Reply:
This key is generated by calling function via 'generate_field_name', which
 is implemented as NONFIELD_CHARS_REGEX.sub('_', name).